### ISSUE TITLE
Error on new wiki page creation with emoji

### ISSUE THREAD

**Issue by jywarren** [publiclab/plots2#2665]
@NiklasJordan said:

> yes I tried that too ;-) Thanks to create this page for me.

> But for this issue I go to https://publiclab.org/wiki/first-contribution

![bildschirmfoto 2018-04-27 um 17 07 51](https://user-images.githubusercontent.com/715033/39369818-8f5b5daa-4a3d-11e8-9a5c-6cfb45934c07.png)

> Then I put my content on it...

![bildschirmfoto 2018-04-27 um 17 08 28](https://user-images.githubusercontent.com/715033/39369848-a2c1d072-4a3d-11e8-83f9-1575a98057b9.png)

> ...press "publish" and get this error page:

![bildschirmfoto 2018-04-27 um 17 09 10](https://user-images.githubusercontent.com/715033/39369888-ba460600-4a3d-11e8-966c-3a43d5ff2bcb.png)

> I am off for the next two hours. But I try to check my phone for your response! Hope this help to find the bug.

****

Let's try to reproduce this -- @Gauravano has already. I'll look through logs but we should try to make a functional test that catches this...

Thanks all!

**Comment by jywarren** [comment]
Maybe if we can get the full text we can try filtering out the sections
that are causing trouble...

**Comment by Gauravano** [comment]
ok, I will try this

**Comment by jywarren** [comment]
@NiklasJordan - can you send us the full text in a Gist using http://gist.github.com so we can try to post it? It's strange but there seem to be characters in the text causing the app to choke... not your fault but we'd like to try to reproduce it to fix it! Thanks for your help with this!

**Comment by NiklasJordan** [comment]
Hej @jywarren, sorry for the late answer. Sure, I've created the gist here: https://gist.github.com/NiklasJordan/da02086827b6f01c0429bacf8cc5ea96

**Comment by jywarren** [comment]
awesome, thank you!!!

**Comment by namangupta01** [comment]
I got it why this is happening. This is happening because of the using of emojis. Emojis take four byte to store data in mysql and is of **utf8mb4** encoding which is a utf8 that take 4 byte. Normally by default it uses 3 byte to store data and use normal utf8 encoding. We can use utf8mb4 for this particular table to support emojis.

**Comment by jywarren** [comment]
Great investigative work! @icarito - is there any downside to changing the
encoding of the revisions table to accommodate emojis?

In the meantime, we could replace the emojis with `:smiley:` format
strings, which are auto-replaced (due to a recent new feature) with the
emoji.

OR we could try to auto-filter emojis and convert them to their `:smiley:`
format strings, and not change the table? Pros, cons?

**Comment by jywarren** [comment]
@icarito any thoughts here -- thank you!

**Comment by icarito** [comment]
Sorry I missed this question! Can this be done in a regular migration? We should definitively test this in staging or unstable instances.

**Comment by jywarren** [comment]
I think it could be. But would you prefer to bypass it by filtering and
replacing with the :____: style emoji? Pros/cons from a storage or
maintenance perspective?

**Comment by icarito** [comment]
I think it makes the most sense to switch to UTF-8, Unicode (utf8_unicode_ci). This should accommodate most languages if not all.

**But** - migrating could break characters if they are not correctly encoded - could lead to mangling some characters at migration time. After migration in `unstable` we should look for pages with international characters, names etc.

Storage and maintenance should be about the same.

**Comment by namangupta01** [comment]
Yes we can do this through migration. If you want i can give it a shot.

**Comment by jywarren** [comment]
@namangupta01  - yes, please - can you try opening a PR for this and we can force push it to `unstable` and  look through a lot of pages to see what happens?

@NiklasJordan i'm sorry this is taking so long! With @Gauravano and @namangupta01 and @icarito's sleuthing, we can now get a non-emoji version up:

https://publiclab.org/first-contribution

If you'd like to make an edit to that page to get credit, that'd be awesome :-)

I'll check back on the last issue now!

**Comment by namangupta01** [comment]
@While trying to reproducing this on my local machine i am getting no error but i am also not getting any text after the emojis in the page i.e all the content after the emojis vanishes.
---Detecting why this is occuring.

**Comment by jywarren** [comment]
https://github.com/railsmachine/utf8mb4_conversion_scripts has a possible solution!

**Comment by jywarren** [comment]
It's pretty thorough, and somewhat involved. I think we should look into it because it may also solve #2209 

@icarito, after the memory-exhaustion issue in #2824 , would you mind looking at this and attempting a migration of `unstable` to see how hard this would be?

**Comment by namangupta01** [comment]
Hi @jywarren I am working today on this will create a pr for this for comment today.

**Comment by jywarren** [comment]
Thinking that this may be fixed in https://github.com/publiclab/plots2/pull/3007 and/or with a follow-up applying this to nodes.
